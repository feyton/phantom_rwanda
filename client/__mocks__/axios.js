const mockAxios = jest.genMockFromModule('axios');
mockAxios.create = jest.fn(() => mockAxios);
mockAxios.get = jest.fn(() =>
  Promise.resolve({
    data: { data: [] }
  })
);
export default mockAxios;
