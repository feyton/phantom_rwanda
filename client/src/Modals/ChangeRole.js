import React, { useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';
import { PrimaryButton } from '../components/Button.js';

function ChangeRole() {
  const navigate = useNavigate();
  const location = useLocation();
  const [newRole, setNewRole] = useState(null);
  const { name, email, role } = location?.state;
  const roles = [
    { name: 'Admin', info: 'Highest priviledges', value: 1 },
    { name: 'Operator', info: 'Can add and view drivers', value: 2 },
    { name: 'Driver', info: 'Can start and stop car', value: 3 }
  ];
  const handleSubmit = (e) => {
    e.preventDefault();
    const { value } = e.target.permission;
    const role = roles.filter((entry) => {
      return entry.value == value;
    });
    setNewRole(role[0]);
    toast('Successfully change the user role to ', { type: 'success' });
    navigate(-1);
  };
  return (
    <div className="font-raleway py-2 px-4">
      <h1 className="font-bold font-raleway text-center text-2xl mb-2">
        Change user's role
      </h1>
      <hr />
      <div className="mt-2">
        <ul>
          <li>
            <b>Name</b>: {name}
          </li>
          <li>
            <b>Email</b>: {email}
          </li>
          <li>
            <b>Role</b>: {newRole?.name || 'Driver'}
          </li>
          <li>
            {newRole?.info && (
              <>
                <b>Priviledges:</b>
                <span>{newRole.info}</span>
              </>
            )}
          </li>
        </ul>
      </div>
      <form
        action="#none"
        id="role-form"
        className="block"
        onSubmit={handleSubmit}
      >
        <label htmlFor="permissions" className="required block my-3">
          Change role:
        </label>
        <select
          required
          aria-required
          name="permission"
          id="permissions"
          className="bg-gray-300 block my-2 border px-2 py-1 rounded-sm placeholder:font-raleway uppercase"
        >
          <option value="" hidden>
            Choose new role
          </option>
          {roles.map((role) => {
            return (
              <option value={role.value} key={role.name}>
                {role.name}
              </option>
            );
          })}
        </select>
        <PrimaryButton type="submit" name="Continue" />
      </form>
    </div>
  );
}

export default ChangeRole;
