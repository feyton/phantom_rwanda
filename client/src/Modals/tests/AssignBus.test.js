import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import AssignModal from '../AssignBus.js';
import ChangeRole from '../ChangeRole.js';

jest.mock('axios');
const mockUseLocationValue = {
  pathname: '/testroute',
  search: '',
  hash: '',
  state: {
    name: 'Fabrice',
    email: 'emanuel@me.com',
    id: 1,
    license: '199787200045637884773',
    plate: 'RAB1000'
  }
};
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn().mockImplementation(() => {
    return mockUseLocationValue;
  })
}));

describe('Modal tests', () => {
  it('Should render the assign modal', async () => {
    const handleSubmitMock = jest.fn();
    const handleCloseMock = jest.fn();
    const data = {
      data: [
        {
          busType: 'Minibus',
          company: 'KBS',
          seats: '20',
          plateNumber: 'RAB1234',
          driver: 'Muhire Girbert',
          route: null,
          id: 1
        }
      ]
    };
    axios.get.mockImplementation(() => Promise.resolve(data));
    const elem = render(
      <MemoryRouter>
        <AssignModal
          handleSubmit={handleSubmitMock}
          handleClose={handleCloseMock}
        />
      </MemoryRouter>
    );
    // expect(await screen.findAllByText(/fabrice@me.com/i)).toHaveLength();

    userEvent.type(screen.getByTestId('bus-plate'), 'rab1234');
    await new Promise((r) => setTimeout(r, 500));
    fireEvent.submit(screen.getByTestId('assign-form'));
    await new Promise((r) => setTimeout(r, 2000));
    expect(elem).toMatchSnapshot();
  });
});

describe('Change Roles', () => {
  it('Renders the modal and submit form', async () => {
    const handleSubmitMock = jest.fn();
    const handleCloseMock = jest.fn();
    const SearchPlateMock = jest.fn(() => new Promise.resolve());
    const elem = render(
      <MemoryRouter>
        <ChangeRole
          name="Fabrice"
          email="fabrice@me.com"
          handleSubmit={handleSubmitMock()}
          SearchPlate={SearchPlateMock}
          handleClose={handleCloseMock}
        />
      </MemoryRouter>
    );
    // expect(await screen.findAllByText(/fabrice@me.com/i)).toHaveLength();
    fireEvent.change(screen.getByTestId('permissions'), {
      target: {
        value: '3'
      }
    });
    expect(elem).toMatchSnapshot();
  });
});
// describe('Function', () => {
//   const axiosMock = jest.spyOn(axios, 'patch');
//   axiosMock.mockImplementation(() => Promise.resolve({}));
//   it('Should pass', async () => {
//     const busInfo = {
//       id: 10
//     };
//     const driverInfo = {
//       id: 10,
//       name: 'Fabrice'
//     };
//     const response = await assignBusToDriver(busInfo, driverInfo);
//     expect(false).toEqual(false);
//   });
// });
