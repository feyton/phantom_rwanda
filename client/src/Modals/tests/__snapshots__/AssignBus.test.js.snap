// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Change Roles Renders the modal and submit form 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="font-raleway py-2 px-4"
      >
        <h1
          class="font-bold font-raleway text-center text-2xl mb-2"
        >
          Change user's role
        </h1>
        <hr />
        <div
          class="mt-2"
        >
          <ul>
            <li>
              <b>
                Name
              </b>
              : 
              Fabrice
            </li>
            <li>
              <b>
                Email
              </b>
              : 
              emanuel@me.com
            </li>
            <li>
              <b>
                Role
              </b>
              : 
              Driver
            </li>
            <li />
          </ul>
        </div>
        <form
          action="#none"
          class="block"
          id="role-form"
        >
          <label
            class="required block my-3"
            for="permissions"
          >
            Change role:
          </label>
          <select
            aria-required="true"
            class="bg-gray-300 block my-2 border px-2 py-1 rounded-sm placeholder:font-raleway uppercase"
            id="permissions"
            name="permission"
            required=""
          >
            <option
              hidden=""
              value=""
            >
              Choose new role
            </option>
            <option
              value="1"
            >
              Admin
            </option>
            <option
              value="2"
            >
              Operator
            </option>
            <option
              value="3"
            >
              Driver
            </option>
          </select>
          <button
            class="bg-primary rounded-md text-white font-bold font-raleway py-2 px-3 hover:bg-hover transition-all block hover:transition-all"
            type="submit"
          >
            Continue
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="font-raleway py-2 px-4"
    >
      <h1
        class="font-bold font-raleway text-center text-2xl mb-2"
      >
        Change user's role
      </h1>
      <hr />
      <div
        class="mt-2"
      >
        <ul>
          <li>
            <b>
              Name
            </b>
            : 
            Fabrice
          </li>
          <li>
            <b>
              Email
            </b>
            : 
            emanuel@me.com
          </li>
          <li>
            <b>
              Role
            </b>
            : 
            Driver
          </li>
          <li />
        </ul>
      </div>
      <form
        action="#none"
        class="block"
        id="role-form"
      >
        <label
          class="required block my-3"
          for="permissions"
        >
          Change role:
        </label>
        <select
          aria-required="true"
          class="bg-gray-300 block my-2 border px-2 py-1 rounded-sm placeholder:font-raleway uppercase"
          id="permissions"
          name="permission"
          required=""
        >
          <option
            hidden=""
            value=""
          >
            Choose new role
          </option>
          <option
            value="1"
          >
            Admin
          </option>
          <option
            value="2"
          >
            Operator
          </option>
          <option
            value="3"
          >
            Driver
          </option>
        </select>
        <button
          class="bg-primary rounded-md text-white font-bold font-raleway py-2 px-3 hover:bg-hover transition-all block hover:transition-all"
          type="submit"
        >
          Continue
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Modal tests Should render the assign modal 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="font-raleway w-fit px-8 py-2"
      >
        <h1
          class="font-bold"
        >
          Driver Info
        </h1>
        <hr />
        <ul>
          <li>
            <span>
              Name: 
            </span>
            <b>
              Fabrice
            </b>
          </li>
          <li>
            <span>
              Email: 
            </span>
            <b>
              emanuel@me.com
            </b>
          </li>
          <li>
            <span>
              License: 
            </span>
            <b>
              199787200045637884773
            </b>
          </li>
          <li>
            <span>
              Assigned bus: 
            </span>
            <b>
              none
            </b>
          </li>
        </ul>
        <hr />
        <div>
          <form
            action="#23"
            class="py-2"
            id="assign-form"
          >
            <label
              class="block font-raleway"
              for="plate"
            >
              Bus plate number
            </label>
            <input
              class="bg-gray-300 my-2 border px-2 py-1 rounded-sm placeholder:font-raleway uppercase"
              id="bus-plate"
              name="plate"
              placeholder="RAB1000"
              required=""
              type="text"
            />
            <p
              class="text-rose-500 mb-2"
            >
              Invalid plate number
            </p>
            <button
              class="bg-primary rounded-md text-white font-bold font-raleway py-2 px-3 hover:bg-hover transition-all block hover:transition-all"
              type="submit"
            >
              Assign
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="font-raleway w-fit px-8 py-2"
    >
      <h1
        class="font-bold"
      >
        Driver Info
      </h1>
      <hr />
      <ul>
        <li>
          <span>
            Name: 
          </span>
          <b>
            Fabrice
          </b>
        </li>
        <li>
          <span>
            Email: 
          </span>
          <b>
            emanuel@me.com
          </b>
        </li>
        <li>
          <span>
            License: 
          </span>
          <b>
            199787200045637884773
          </b>
        </li>
        <li>
          <span>
            Assigned bus: 
          </span>
          <b>
            none
          </b>
        </li>
      </ul>
      <hr />
      <div>
        <form
          action="#23"
          class="py-2"
          id="assign-form"
        >
          <label
            class="block font-raleway"
            for="plate"
          >
            Bus plate number
          </label>
          <input
            class="bg-gray-300 my-2 border px-2 py-1 rounded-sm placeholder:font-raleway uppercase"
            id="bus-plate"
            name="plate"
            placeholder="RAB1000"
            required=""
            type="text"
          />
          <p
            class="text-rose-500 mb-2"
          >
            Invalid plate number
          </p>
          <button
            class="bg-primary rounded-md text-white font-bold font-raleway py-2 px-3 hover:bg-hover transition-all block hover:transition-all"
            type="submit"
          >
            Assign
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
